README.md
# Baiduwp-PHP

[ENGLISH](README.md) | [中文](README_ZH.md)

## ⚠️ Disclaimer
Please note that this project is only for learning and research purposes, and the author is not responsible for any legal consequences caused by the use of this project.

This project will not store any files on the server, completely relying on the Baidu Netdisk API interface. If you have question about copyright, please contact Baidu: [https://newcopyright.baidu.com/](https://newcopyright.baidu.com/)

## 📝 Introduction
This project is a PHP version of [baiduwp](htttps://github.com/TkzcM/baiduwp).

It can get the file information of Baidu Netdisk through the API interface and the Cookie (BDUSS) of the SVIP account, and then display it on the web page.

The essence is to use the premium account (SVIP account) to get the download link and send it to the visitor.


## 🖥️ Demo
### Demo Site
- [Click here go to Demo](https://imwcr.cn/api/bdwp/)<br/>

### Video tutorial
- [[YouTube] How to download files from pan.baidu.com without login](https://www.youtube.com/watch?v=8u8jx3Owuc8)

### Screenshot
![浅色及英文模式](https://s2.loli.net/2023/02/04/cs1EtFXpHDPS2AB.png)
![文件列表](https://s2.loli.net/2023/02/04/hL2pDEyHQFb6BKR.png)
![解析详情](https://s2.loli.net/2023/02/04/GZBsmz6xgShjuA2.png)

## 💡Contact
- Author: Yuan_Tuo
- Site: https://imwcr.cn/
- Email: yuantuo666@gmail.com
- Telegram: https://t.me/yuantuo666


## Related Works
Following are some related works with this project. If you want to apply to add new project here, please draft new issue with brief introduction. 
- [alist-org/alist](https://github.com/alist-org/alist) 🗂️A file list/WebDAV program that supports multiple storages, powered by Gin and Solidjs. / 一个支持多存储的文件列表/WebDAV程序，使用 Gin 和 Solidjs。
- [codehub666/94list](https://github.com/codehub666/94list) 百度网盘分享链接分析渲染列表辅助下载开源程序
- [z-mio/baiduwp-bot](https://github.com/z-mio/baiduwp-bot) 一个基于baiduwp-php API的百度网盘解析bot
- [monkeyWie/gopeed-extension-baiduwp](https://github.com/monkeyWie/gopeed-extension-baiduwp) Gopeed 百度网盘下载扩展。


## 🔔Thanks
- [baiduwp JavaScript 版](https://github.com/TkzcM/baiduwp "baiduwp 项目")
- [PanDownload 网站](https://pandownload.com/ "PanDownload 网站")
- [Bootstrap 深色模式](https://github.com/vinorodrigues/bootstrap-dark "bootstrap-dark 项目")
- [ThinkPHP](https://github.com/top-think/think)

docs/CHANGELOG.md
# Update
### `4.0.3` 版本：
- 更新日期：2023-11-24
- 修改内容：
  - ⚠错误修复
    - 修复无法自动更换帐号问题
    - 增加错误提示

### `4.0.2` 版本：
- 更新日期：2023-11-03
- 修改内容：
  - ⚠错误修复
    - 修复无法解析 4G 以上文件
  - 新功能
    - 支持 Apache 环境下的子目录和伪静态自动配置（Nginx 仍然需要手动设置）

### `4.0.1` 版本：
- 更新日期：2023-09-28
- 修改内容：
  - ⚠错误修复
    - 修复 docker 无法自动创建数据库的
    - 修复首页账号状态检测总是显示限速
  - ♻代码优化
    - 更新 API 文档

### `4.0.0` 版本：
- 更新日期：2023-09-24
- 修改内容：
  - 💥新增功能
    - 支持 SQLite 数据库
  - ♻代码优化
    - 使用 ThinkPHP 重构

### `3.0.2` 版本：
- 更新日期：2023-04-07
- 修改内容：
  - ⚠错误修复
    - 修复限速账号检查
    - 修复不安全网站无法复制下载地址
    - 修复后台无法获取SVIP账号状态
  - ♻代码优化
    - 改进后台管理错误提示
    - 改进翻译函数
    - 后端账号状态分离至API接口
    - 代码中版权信息改为项目地址

### `3.0.1` 版本：
- 更新日期：2023-04-05
- 修改内容：
  - ⚠错误修复
    - 修复根目录下压缩包被识别为文件夹的问题 #255 #256 #257

### `3.0.0` 版本：
- 更新日期：2023-04-04
- 修改内容：
  - 💥新增功能
    - 新增后端API接口及文档 #237
    - 后台获取账号状态增加缓存 #253
    - 更新UI
  - ♻代码优化
    - 重构解析有关操作为类

### `2.2.7` 版本：
- 更新日期：2023-03-15
- 修改内容：
  - 💥新增功能
    - 支持 docker 部署
  - ⚠错误修复
    - fix 未开启数据库无法下载 #251
  - ♻代码优化
    - 运行时检查 Cookie 设置
    - 优化错误提示

### `2.2.6` 版本：
- 更新日期：2023-03-09
- 修改内容：
  - ⚠错误修复
    - fix 更新提示“不再提示”逻辑 #197
    - fix 一直提示升级2.2.6 #244
  - ♻代码优化
    - 安装时自动检查Cookie设置
    - 链接失效自动重新生成sign
    - DEBUG模式相关信息显示在控制台中
    - 删除无用代码
    - 格式化部分代码

### `2.2.5` 版本：
- 更新日期：2022-11-24
- 修改内容：
  - ⚠错误修复
    - 修复 下载报错 9019 问题 #225 #227

### `2.2.4` 版本：
- 更新日期：2022-10-24
- 修改内容：
  - ⚠错误修复
    - 修复 因百度对下载接口调整导致的失效问题 #224
  - ♻代码优化
    - 可发送下载任务至文件蜈蚣下载 #187

### `2.2.3` 版本：
- 更新日期：2022-10-07
- 修改内容：
  - ⚠错误修复
    - 修复 因百度 09.28 对sign接口调整导致的失效问题 #219 #221
  - ♻代码优化
    - 支持识别新版本分享链接 #216
    - 提取码错误提示完善 #214 #217 #218
### `2.2.2` 版本：
- 更新日期：2022-09-22
- 修改内容：
  - ⚠错误修复
    - 完成 微信API 解析模式
    - 修复 因百度 09.21 调整导致的解析文件夹 失效问题

### `2.2.1` 版本：
- 更新日期：2022-09-17
- 修改内容：
  - ⚠错误修复
    - 修复 [#191 Possible SQLI and XSS vulnerabilities](https://github.com/yuantuo666/baiduwp-php/issues/191)
    - 修复 [#160 【BUG 反馈】文件名中的 ' 没有转义](https://github.com/yuantuo666/baiduwp-php/issues/160)
  - ♻代码优化
    - 代码结构调整

### `2.2.0` 版本：
- 更新日期：2021-09-21
- 修改内容：
  - 💥新增功能
    - ✨支持老版本链接兼容加载
    - ✨会员账号批量导入支持设置 STOKEN 以及独自用户名
      - 格式：BDUSS----STOKEN----账号名称
    - 增加默认语言选项
    - 更新提示增加展开动画，且可以取消提示
    - 语言切换支持苹果 Safari 浏览器 zh-cn 标识
    - 增加安装提示
  - ⚠错误修复
    - 修复加载或切换页面白屏问题
    - 修复深色模式下首页输入框高度错误
    - settings.php 以及 install.php 页面更新提示与 ready.js 同步
    - 修复二维码内容无法加载问题
    - 安装页面设置站点默认语言 bug 修复
  - ♻代码优化
    - 请求静态文件时会带上版本号，减少更新时的因缓存导致的bug
    - 小屏幕不展示二维码
      - 手机端展示二维码时会挡住下载链接，导致无法复制
    - 获取文件失败提示内容修改

### `2.1.9.1` 版本：
- 更新日期：2021-07-22
- 修改内容：
  - 安全性更新
    - 修复 API SQL 注入问题 #148
  - 问题修复
    - 修复 `install` 和 `settings` 页面更新提示有误的问题
    - 修复 `install` 写入 HTTP 响应正文后仍写入响应头的 bug（取消了自动跳转功能）

### `2.1.9` 版本：
- 更新日期：2021-07-14
- 修改内容：
  - 新增功能
    - 下载链接生成二维码 #123
    - 自动检查更新
  - 功能优化
    - 在跟随浏览器模式下，颜色模式随浏览器设置实时更新
    - 优化颜色模式逻辑
    - 优化用户设置页面前端逻辑
    - 优化二维码生成
    - 优化样式表加载
    - 优化 JS 逻辑和加载
  - 错误修复
    - 修复了正则表达式有误导致的子域名无法正常使用 Aira2 下载问题 #137
  - 其它
    - 更新 .gitignore
    - 格式化代码

### `2.1.8` 版本：
- 更新日期：2021-06-27
- 修改内容：
  - 💥新增功能
    - 增加对 IP 段的黑白名单设置
  - ⚠错误修复
    - 修复带单引号的文件名写入数据库时出错的问题 #131
    - 修复用户界面显示的解析链接有效时间并不会随设置变化的问题 #132

### `2.1.7` 版本：
- 更新日期：2021-05-18
- 修改内容：
  - 💥新增功能
    - 新增aria2安卓端
    - 新增解析链接失效时间设置
    - 关于页面自动生成二维码
  - ⚠错误修复
    - 修复语言设置错误提示中转跳用户设置页面链接错误
    - 修复手机端推送 aria2 因文件夹不存在出错
  - ♻代码优化
    - 在新标签页打开Motrix首页
    - ws(s)链接检测支持IPv6

### `2.1.6` 版本：
- 更新日期：2021-05-01
- 修改内容：
  - 💥新增功能
    - 新增Motrix下载方式
  - ⚠错误修复
    - 修复aria2推送token问题
  - ♻代码优化
    - 隐藏未启用数据库时奇怪的提示
    - 增加推送错误提示
    - 处理一些文字格式
    - ws(s)链接时检测是否有效
    - aira存储信息方式修改为localStorage

### `2.1.5` 版本：
- 更新日期：2021-04-03
- 修改内容：
  - 💥新增功能
    - 账号状态检查
    - Aria2在线管理页面
  - ⚠错误修复
    - 修复未开启数据库情况下无法正常获取文件下载地址   GitHub: @kwxiaozhu
    - 修复GetDir()支持PHP8  GitHub: @zzjin
    - 改用WebSocket推送下载链接到aria2
  - ♻代码优化
    - 增加账号失效提示
    - 版权信息改为Github项目页面
    - 删除不必要的版本检测和版本号

### `2.1.4` 版本：
- 更新日期：2021-03-28
- 修改内容：
  - 💥新增功能
    - 支持通过微信API获取50MB以下文件
  - ⚠错误修复
    - 修复由于百度网盘更新页面js代码引起的项目失效

### `2.1.3` 版本：
- 更新日期：2021-03-14
- 修改内容：
  - 💥新增功能
    - 兼容老方法获取randsk(BDCLND) #84
    - aria2推送时附上文件名 #78 GitHub: @kwxiaozhu
  - ⚠错误修复
    - 安装页面未处理默认情况丢失前缀 #76
  - ♻代码优化
    - 首页安装后状态提示完善 #76 #81
    - 根页面获取失败提示完善
    - 账号拉黑提示 #83 #84 #86

###  `2.1.2` 版本：
- 更新日期：2021-02-20
- 备注：2.1.0版本和2.1.1版本存在安装问题，请勿安装
- 修改内容：
  - ⚠错误修复
    - 修复无法安装问题

### [不稳定] `2.1.1` 版本：
- 更新日期：2021-02-18
- 修改内容：
  - ⚠错误修复
    - 修复数据库中会员账号失效后一直刷新页面问题
    - 修复旧版本无法正常升级问题

## [不稳定] `2.1.0` 版本：
- 更新日期：2021-02-17
<!-- 同志们，写更新日志要细致啊，不要写笼统的！ -->
- 修改内容：
  - 💥新增功能
    - 安装程序 `install.php` 自动检测旧版本配置文件 `config.php` 是否存在，若存在自动导入旧版本配置
    - 增加选择是否取消下载次数提醒功能
    - ✨安装时支持保留数据库数据
    - ✨后台管理页面支持删除数据
    - ✨增加四种SVIP账号切换模式
    - 增加首页公告自定义功能
  - 💪安全增强
    - 安装程序 `install.php` 自动检测是否安装过，如果安装则需进入管理员页面登录
  - ⚠错误修复
    - 修复部分页面检查密码功能失效问题
    - 修复首页小圆点无颜色错误
    - 修复不支持色彩模式的浏览器无法显示 `Sweetalert2` 弹窗问题
    - 修复解析数据一直为 `2.00GB` 问题
    - 修复管理员密码错误不提示
  - ♻代码优化
    - ✨将 `settings.php` 内部分请求方式改为 `ajax` ，增加加载提示框 <!-- 搞了四个小时，累死 -->
    - 优化提示文本（语法、严谨程度等），给一些提示框增加图标
    - 增加部分配置异常地处理程序
    - 优化部分 PHP 和 JavaScript 代码

## [后台管理系统] `1.4.5`：
- 更新日期：2021-01-25
- 修改内容
  - 增加后台系统
  - 增加数据统计功能
  - 增加限速账号自动切换功能
  - 增加API接口文件
  - 增加解析记录查询
  - IP黑白名单
  - 下载次数限制修改
-  [Commit记录](https://github.com/yuantuo666/baiduwp-php/commit/a2f76c9d9f4c70d349279631d0d0dba01cee07ef) （1,169 additions and 62 deletions.）

http://www.dupan.cc/ （站长QQ33703259） 所发布的源码已被篡改，添加了后台并加密，添加的bduss会被上传网站后台，请勿下载使用

## [稳定] `1.4.3` 版本：
- 更新日期：2020-10-21
- 修改内容
  - 后台增加MySQL数据库，保存8小时内解析文件。
  - 限制同一IP及设备的解析次数。

### `1.4.2` 版本：
- 更新日期：2020-08-29
- 修改内容
  - 列表页面新增超时提醒，5min后弹窗提示。
  - 修复在线播放功能，在设置UA情况下可以播放50MB以上文件。
  - 优化代码，删除打开文件夹每次查询密码是否正确代码。
  - 加入运行时间计算，在控制台中可以查看。
  - 将SVIP的BDUSS分离开，便于后期维护。
  - 隐藏旧链接显示的sharelinkXXX-XXX文件夹（此文件夹无法正常打开）。
  - 增加调试模式，便于反馈问题。
  - 增加自动从分享文本中提取验证密码功能。

### `1.4.1` 版本：
- 更新日期：2020-08-27
- 修改内容
  - 修改POST内容，让调用接口暂时失效
  - 增加直链解析，可以不设置UA下载（不过并不稳定，且只支持50MB以下文件）

### `1.4.0` 版本：
- 更新日期：2020-08-27
- 修改内容
  - 增加推送至aria2功能
  - 增加面包屑导航，便于寻找指定文件夹

app/Parse.php
<?php

namespace app;

use app\Req;
use app\Tool;
use think\facade\Db;

/**
 * PanDownload 网页复刻版，PHP 语言版
 *
 * @author Yuan_Tuo <yuantuo666@gmail.com>
 * @link https://github.com/yuantuo666/baiduwp-php
 */
class Parse
{
	public static function getSign(string $surl = "", string $share_id = "", string $uk = ""): array
	{
		// construct url
		$params = "";
		if (substr($surl, 0, 1) == "2") {
			$surl_2 = substr($surl, 1);
			$surl_2 = base64_decode($surl_2);
			[$uk, $share_id] = explode("&", $surl_2);
			$surl = "";
		}
		if ($surl) $params .= "&surl=$surl";
		if ($share_id) $params .= "&shareid=$share_id";
		if ($uk) $params .= "&uk=$uk";
		$url = "https://pan.baidu.com/share/tplconfig?$params&fields=sign,timestamp&channel=chunlei&web=1&app_id=250528&clienttype=0";
		$header = [
			"User-Agent: netdisk;pan.baidu.com",
			"Cookie: " . config('baiduwp.cookie'),
		];
		$result = Req::GET($url, $header);
		$result = json_decode($result, true, 512, JSON_BIGINT_AS_STRING);
		if (($result["errno"] ?? 1) == 0) {
			$sign = $result["data"]["sign"];
			$timestamp = $result["data"]["timestamp"];
			return [0, $sign, $timestamp];
		} else {
			return [-1, $result["show_msg"] ?? "", ""];
		}
	}

	public static function decodeSceKey($seckey)
	{
		$seckey = str_replace("-", "+", $seckey);
		$seckey = str_replace("~", "=", $seckey);
		return str_replace("_", "/", $seckey);
	}

	public static function decryptMd5($md5)
	{
		if (preg_match('/^.{9}[a-f0-9]/', $md5) && ctype_xdigit(substr($md5, 9, 1))) {
			return $md5;
		}
		$key = dechex(ord(substr($md5, 9, 1)) - ord('g'));
		$key2 = substr($md5, 0, 9) . $key . substr($md5, 10, strlen($md5));
		$key3 = "";
		for ($a = 0; $a < strlen($key2); $a++) {
			$key3 .= dechex(hexdec($key2[$a]) ^ (15 & $a));
		}
		return substr($key3, 8, 8) . substr($key3, 0, 8) . substr($key3, 24, 8) . substr($key3, 16, 8);
	}

	public static function getList($surl, $pwd, $dir, $sign = "", $timestamp = ""): array
	{
		$message = [];
		if (!$sign || !$timestamp) {
			list($status, $sign, $timestamp) = self::getSign($surl);
			if ($status !== 0) {
				$sign = '';
				$timestamp = '1';
				$message[] = "无传入，自动获取sign和timestamp失败, $sign";
			} else {
				$message[] = "无传入，自动获取sign和timestamp成功: $sign, $timestamp";
			}
		}

		$IsRoot = $dir == "";
		$file_list = [];
		$Page = 1;
		// 获取所有文件 fix #86
		while (true) {
			$Filejson = self::getListApi($surl, $dir, $IsRoot, $pwd, $Page);
			if (config('app.debug')) $message[] = json_encode($Filejson);
			if ($Filejson["errno"] ?? 999 !== 0) {
				return self::listError($Filejson, $message);
			}
			foreach ($Filejson['data']['list'] as $v) {
				$file_list[] = $v;
			}
			if (count($Filejson['data']["list"]) < 1000) break;
			$Page++;
		}
		$randSk = urlencode(self::decodeSceKey($Filejson["data"]["seckey"]));
		$shareid = $Filejson["data"]["shareid"];
		$uk = $Filejson["data"]["uk"];

		// breadcrumb
		$DirSrc = [];
		if (!$IsRoot) {
			$Dir_list = explode("/", $dir);

			for ($i = 1; $i <= count($Dir_list) - 2; $i++) {
				if ($i == 1 and strstr($Dir_list[$i], "sharelink")) continue;
				$fullsrc = strstr($dir, $Dir_list[$i], true) . $Dir_list[$i];
				$DirSrc[] = array("isactive" => 0, "fullsrc" => $fullsrc, "dirname" => $Dir_list[$i]);
			}
			$DirSrc[] = array("isactive" => 1, "fullsrc" => $dir, "dirname" => $Dir_list[$i]);
		}
		$Filenum = count($file_list);
		$FileData = [];
		$RootData = array(
			"src" => $DirSrc,
			"timestamp" => $timestamp,
			"sign" => $sign,
			"randsk" => $randSk,
			"shareid" => $shareid,
			"surl" => $surl,
			"pwd" => $pwd,
			"uk" => $uk,
		);

		foreach ($file_list as $file) {
			if ($file["isdir"] == 0) { // 根目录返回的居然是字符串 #255
				//文件
				$FileData[] = array(
					"isdir" => 0,
					"name" => $file["server_filename"],
					"fs_id" => number_format($file["fs_id"], 0, '', ''),
					"size" => $file["size"],
					"uploadtime" => $file["server_ctime"],
					"md5" => $file["md5"],
					"dlink" => $file["dlink"]
				);
			} else {
				//文件夹
				$FileData[] = array(
					"isdir" => 1,
					"name" => $file["server_filename"],
					"path" => $file["path"],
					"size" => $file["size"],
					"uploadtime" => $file["server_ctime"]
				);
			}
		}

		return array(
			"error" => 0,
			"isroot" => $IsRoot,
			"dirdata" => $RootData,
			"filenum" => $Filenum,
			"filedata" => $FileData,
			"message" => $message
		);
	}

	public static function download($fs_id, $timestamp, $sign, $randsk, $share_id, $uk)
	{
		if (!$fs_id || !$randsk || !$share_id || !$uk) {
			return [
				'error' => -1,
				'msg' => '参数错误',
			];
		}
		$message = [];

		$ip = Tool::getIP();
		$isipwhite = FALSE;
		if (config('baiduwp.db')) {
			$data = Db::connect()->table("ip")->where('ip', $ip)->find();
			if ($data) {
				// 存在 判断类型
				if ($data->type === -1) {
					// 黑名单
					return array("error" => -1, "msg" => "当前ip已被加入黑名单，请联系站长解封", "ip" => $ip);
				} elseif ($data->type === 0) {
					// 白名单
					$message[] = "当前ip为白名单~ $ip";
					$isipwhite = TRUE;
				}
			}
		}

		// check if the timestamp is valid
		if (time() - $timestamp > 300) {
			// try to get the timestamp and sign
			list($_status, $sign, $timestamp) = self::getSign("", $share_id, $uk);
			if ($_status !== 0) {
				$message[] = "超时，自动获取sign和timestamp失败, $sign";
			} else {
				$message[] = "超时，自动获取sign和timestamp成功: $sign, $timestamp";
			}
		}

		$json4 = self::getDlink($fs_id, $timestamp, $sign, $randsk, $share_id, $uk);
		$errno = $json4["errno"] ?? 999;
		if ($errno !== 0) {
			if (config('app.debug')) {
				$message[] = "获取下载链接失败: " . json_encode($json4);
			} else {
				$message[] = "获取下载链接失败: " . ($json4["error_msg"] ?? "未知错误");
			}
			return self::downloadError($json4, $message);
		}

		$dlink = $json4["list"][0]["dlink"] ?? "";
		// 获取文件相关信息
		$md5 = $json4["list"][0]["md5"];
		$md5 = self::decryptMd5($md5); // 尝试解密md5
		$filename = $json4["list"][0]["server_filename"] ?? "";
		$size = $json4["list"][0]["size"] ?? "0";
		$path = $json4["list"][0]["path"] ?? "";
		$server_ctime = (int)$json4["list"][0]["server_ctime"] ?? 28800; // 服务器创建时间 +8:00

		$FileData = array(
			"filename" => $filename,
			"size" => $size,
			"path" => $path,
			"uploadtime" => $server_ctime,
			"md5" => $md5
		);

		if ($size <= 5242880) { // 5MB
			return array("error" => 0, "filedata" => $FileData, "directlink" => $dlink, "user_agent" => "LogStatistic", "message" => $message);
		}

		list($ID, $cookie) = ["-1", config('baiduwp.cookie')];

		if (config('baiduwp.db')) {
			$link_expired_time = config('baiduwp.link_expired_time') ?? 8;

			// 查询数据库中是否存在已经保存的数据
			$data = Db::connect()
				->table("records")
				->where('md5', $md5)
				->where('time', '>', date('Y-m-d H:i:s', time() - (int)($link_expired_time * 3600)))
				->find();

			if ($data) {
				// 存在
				$realLink = $data['link'];
				return array("error" => 0, "usingcache" => true, "filedata" => $FileData, "directlink" => $realLink, "user_agent" => "LogStatistic", "message" => $message);
			}

			// 判断今天内是否获取过文件
			if (!$isipwhite) { // 白名单跳过
				// 获取解析次数
				$count = Db::connect()->table('records')->where('ip', $ip)
					->where('time', '>', date('Y-m-d 00:00:00', time()))
					->count();

				if ($count > config('baiduwp.times')) {
					// 提示无权继续
					return array("error" => 1, "msg" => "今日解析次数已达上限，请明天再试", "ip" => $ip);
				}
				// 获取解析流量
				$flow = Db::connect()->table('records')->where('ip', $ip)
					->where('time', '>', date('Y-m-d 00:00:00', time()))
					->sum('size');
				if ($flow > config('baiduwp.flow') * 1024 * 1024 * 1024) {
					// 提示无权继续
					return array("error" => 1, "msg" => "今日解析流量已达上限，请明天再试", "ip" => $ip);
				}
			}

			$query = Db::connect()->table("account")->where('status', 0);
			if (config('baiduwp.random_account')) {
				$query = $query->order('last_used_at', 'asc');
			} else {
				$query = $query->order('created_at', 'asc');
			}
			$data = $query->find();
			if ($data) {
				list($ID, $cookie) = [$data['id'], $data['cookie']];
				// 更新最后使用时间
				Db::connect()->table('account')->where('id', $ID)->update(['last_used_at' => date('Y-m-d H:i:s', time())]);
			}
		}

		$SVIP_BDUSS = Tool::getSubstr($cookie, 'BDUSS=', ';');

		// 开始获取真实链接
		$headerArray = array('User-Agent: LogStatistic', 'Cookie: BDUSS=' . $SVIP_BDUSS . ';');

		$header = Req::HEAD($dlink, $headerArray); // 禁止重定向
		if (!strstr($header, "Location")) {
			// fail
			$message[] = "获取真实链接失败 $header";
		}
		$header = str_replace("https://", "http://", $header);
		$real_link = Tool::getSubstr($header, "Location: ", "\r\n"); // delete http://

		// 1. 使用 dlink 下载文件   2. dlink 有效期为8小时   3. 必需要设置 User-Agent 字段   4. dlink 存在 HTTP 302 跳转
		if (!$real_link || strlen($real_link) < 20) {
			$body = Req::GET($dlink, $headerArray);
			$body_decode = json_decode($body, true);

			$message[] = "获取真实下载链接失败：" . json_encode($body_decode);

			if (config('baiduwp.db') && config('baiduwp.check_speed_limit') && $ID != "-1") {
				$result = Db::connect()->table('account')->where('id', $ID)->update(['status' => -1]);
				if ($result) {
					return array("error" => -1, "msg" => "SVIP账号自动切换成功，请重新请求获取下载地址", "message" => $message);
				} else {
					return array("error" => -1, "msg" => "SVIP账号自动切换失败", "message" => $message);
				}
			}
			return self::realLinkError($body_decode, $message); // 获取真实链接失败
		}
		if (str_contains($real_link, "qdall01") || !str_contains($real_link, 'tsl=0')) {
			// 账号限速
			$message[] = "SVIP账号限速";
			if (config('baiduwp.db') && config('baiduwp.check_speed_limit') && $ID != "-1") {
				$result = Db::connect()->table('account')->where('id', $ID)->update(['status' => -1]);
				if ($result) {
					return array("error" => -1, "msg" => "SVIP账号自动切换成功，请重新请求获取下载地址", "message" => $message);
				} else {
					return array("error" => -1, "msg" => "SVIP账号自动切换失败", "message" => $message);
				}
			}
		}

		// 记录下使用者ip，下次进入时提示
		if (config('baiduwp.db')) {
			$result = Db::connect()->table("records")->insert([
				'ip' => $ip,
				'ua' => request()->header('User-Agent'),
				'name' => $filename,
				'size' => $size,
				'md5' => $md5,
				'link' => $real_link,
				'time' => date('Y-m-d H:i:s', time()),
				'account' => $ID
			]);

			if (!$result) {
				// 保存错误
				return array("error" => -1, "msg" => "数据库保存错误", "message" => $message);
			}
		}

		return array("error" => 0, "filedata" => $FileData, "directlink" => $real_link, "user_agent" => "LogStatistic", "message" => $message);
	}

	private static function getListApi(string $Shorturl, string $Dir, bool $IsRoot, string $Password, int $Page = 1)
	{
		$Url = 'https://pan.baidu.com/share/wxlist?channel=weixin&version=2.2.2&clienttype=25&web=1';
		$Root = ($IsRoot) ? "1" : "0";
		$Dir = urlencode($Dir);
		if (substr($Shorturl, 0, 1) == "2") {
			$Shorturl = substr($Shorturl, 1);
			$Shorturl = base64_decode($Shorturl);
			[$uk, $share_id] = explode("&", $Shorturl);
			$params = "&uk=$uk&shareid=$share_id";
		} else {
			$params = "&shorturl=$Shorturl";
		}
		$Data = "$params&dir=$Dir&root=$Root&pwd=$Password&page=$Page&num=1000&order=time";
		$BDUSS = Tool::getSubstr(config('baiduwp.cookie'), 'BDUSS=', ';');
		$header = ["User-Agent: netdisk", "Cookie: BDUSS=$BDUSS", "Referer: https://pan.baidu.com/disk/home"];
		return json_decode(Req::POST($Url, $Data, $header), true);
	}

	private static function getDlink(string $fs_id, string $timestamp, string $sign, string $randsk, string $share_id, string $uk, int $app_id = 250528)
	{ // 获取下载链接
		$url = 'https://pan.baidu.com/api/sharedownload?app_id=' . $app_id . '&channel=chunlei&clienttype=12&sign=' . $sign . '&timestamp=' . $timestamp . '&web=1'; // 获取下载链接

		if (strstr($randsk, "%")) $randsk = urldecode($randsk);
		$data = "encrypt=0" . "&extra=" . urlencode('{"sekey":"' . $randsk . '"}') . "&fid_list=[$fs_id]" . "&primaryid=$share_id" . "&uk=$uk" . "&product=share&type=nolimit";
		$header = array(
			"User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36 Edg/110.0.1587.69",
			"Cookie: " . config('baiduwp.cookie'),
			"Referer: https://pan.baidu.com/disk/home"
		);
		return json_decode(Req::POST($url, $data, $header), true);
		// 没有 sekey 参数就 118, -20出现验证码
	}

	private static function listError($Filejson, $message): array
	{
		if (empty($Filejson)) {
			return [
				"error" => -1,
				"title" => "获取到的列表为空",
				"msg" => "请确保已正确设置 Cookie 且服务器网络连接正常",
				"message" => $message
			];
		}
		// 解析异常
		$ErrorCode = $Filejson["errtype"] ?? ($Filejson["errno"] ?? 999);
		$ErrorMessage = [
			"mis_105" => "你所解析的文件不存在~",
			"mispw_9" => "提取码错误",
			"mispwd-9" => "提取码错误",
			"mis_2" => "不存在此目录",
			"mis_4" => "不存在此目录",
			5 => "不存在此分享链接或提取码错误",
			3 => "此链接分享内容可能因为涉及侵权、色情、反动、低俗等信息，无法访问！",
			0 => "啊哦，你来晚了，分享的文件已经被删除了，下次要早点哟。",
			10 => "啊哦，来晚了，该分享文件已过期",
			8001 => "普通账号可能被限制，请检查普通账号状态",
			9013 => "普通账号被限制，请检查普通账号状态",
			9019 => "普通账号 Cookie 状态异常，请检查：账号是否被限制、Cookie 是否过期（前往网站设置页面修改）",
			999 => "错误 -> " . json_encode($Filejson)
		];
		return [
			"error" => -1,
			"title" => "获取列表错误 ($ErrorCode)",
			"msg" => $ErrorMessage[$ErrorCode] ?? "未知错误，如多次出现请向提出issue反馈",
			"message" => $message
		];
	}

	private static function downloadError($json4, $message): array
	{
		$errno = $json4["errno"] ?? 999;
		$error = [
			999 => ["请求错误", "请求百度网盘服务器出错，请检查网络连接或重试"],
			-20 => ["触发验证码(-20)", "请等待一段时间，再返回首页重新解析。"],
			-9 => ["文件不存在(-9)", "请返回首页重新解析。"],
			-6 => ["账号未登录(-6)", "请检查普通账号是否正常登录。"],
			-1 => ["文件违规(-1)", "您下载的内容中包含违规信息"],
			2 => ["下载失败(2)", "下载失败，请稍候重试"],
			112 => ["链接超时(112)", "获取链接超时，每次解析列表后只有5min有效时间，请返回首页重新解析。"],
			113 => ["传参错误(113)", "获取失败，请检查参数是否正确。"],
			116 => ["链接错误(116)", "该分享不存在"],
			118 => ["没有下载权限(118)", "没有下载权限，请求百度服务器时，未传入sekey参数或参数错误。"],
			110 => ["服务器错误(110)", "服务器错误，可能服务器IP被百度封禁，请切换 IP 或更换服务器重试。"],
			121 => ["服务器错误(121)", "你选择操作的文件过多，减点试试吧"],
			8001 => ["普通账号错误(8001)", "普通账号可能被限制，请检查普通账号状态"],
			9013 => ["普通账号错误(9013)", "普通账号被限制，请检查普通账号状态"],
			9019 => ["普通账号错误(9019)", "普通账号 Cookie 状态异常，请检查：账号是否被限制、Cookie 是否过期（前往网站设置页面修改）"],
		];

		if (isset($error[$errno])) return [
			"error" => -1,
			"title" => $error[$errno][0],
			"msg" => $error[$errno][1],
			"message" => $message
		];
		else return [
			"error" => -1,
			"title" => "获取下载链接失败 ($errno)",
			"msg" => "未知错误！错误：" . json_encode($json4),
			"message" => $message
		];
	}

	private static function realLinkError($body_decode, $message): array
	{
		$ErrorCode = $body_decode["errno"] ?? ($body_decode["error_code"] ?? 999);
		$ErrorMessage = [
			8001 => "SVIP 账号可能被限制，请检查 SVIP 的 Cookie 是否设置正确且有效",
			9013 => "SVIP 账号被限制，请检查更换 SVIP 账号",
			9019 => "SVIP 账号可能被限制，请检查 SVIP 的 Cookie 是否设置正确且有效",
			31360 => "下载链接超时，请刷新页面重试。若重试后仍报错，请检查普通帐号 Cookie 是否过期",
			31362 => "下载链接签名错误，请检查 UA 是否正确",
			999 => "错误 -> " . json_encode($body_decode)
		];

		return [
			"error" => -1,
			"title" => "获取下载链接失败 ($ErrorCode)",
			"msg" => $ErrorMessage[$ErrorCode] ?? "未知错误！错误：" . json_encode($body_decode),
			"message" => $message
		];
	}
}
